UDP-70 - Send both the vanity url and the full url to the scoring API.

wrandall22 opened the pull request:
This PR sends both the vanity URL and the full URL to the scoring API. Before this code, it would send only the vanity URL if it had one, and otherwise send the full URL. Unfortunately, this gives us incomplete data. Specifically, there are certain pages that have vanity URLs that can also be reached by their full URLs (not all of them fall in this category: if I remember correctly this is determined by the redirectVanityUrl property).

In order to not lose any scoring data and to not make the code here overly complex, we are sending both the full and vanity URLs (if the page has one) every time.

[JIRA](https://jira.cru.org/browse/UDP-70)

wrandall22 committed:
Send both the vanity url and the full url to the scoring API.

mattdrees commented on core/src/test/java/org/cru/contentscoring/core/service/impl/ContentScoreUpdateServiceImplTest.java:
```diff
@@ -166,10 +166,31 @@ public void testGetPageUrl() {
+
+        String vanityUrl = updateService.getPageUrl(page, false);
+        assertThat(vanityUrl, is(nullValue()));
+    }
```
I'm confused by the scenario this test represents. The PR says that you are trying to send "both" the vanity url and the full url, but in this scenario the real URL is not sent, right?

wrandall22 replied:
You're right, I had it that way at first, but second guessed myself at some point. It should send real URL in this case.

mattdrees reviewed (commented):
not sure yet if I'm ok with the code, because something about the logic here is not making sense to me

wrandall22 committed:
Fix logic and improve tests

mattdrees commented on core/src/test/java/org/cru/contentscoring/core/service/impl/ContentScoreUpdateServiceImplTest.java:
```diff
@@ -150,33 +155,128 @@ public void testSetContentScoreUpdateDateRetry() throws RepositoryException {
+        }
+
+        assertThat(correctPaths, is(equalTo(2)));
+    }
```
So this test sends:
 * `https://vanity.com/content/test/us/en/vanity-url.html`
 * `https://vanity.com/vanity-url`

The latter makes sense to me, but I thought the former should have been `https://vanity.com/content/test/us/en/page-path.html`. Am I wrong, and we actually want vanity-url.html? (I am admittedly not very knowledgeable about how vanity urls work.)

wrandall22 replied:
So, it has to do with how the externalizer works. I ran tests through again to see it was doing what I had thought when I first wrote these tests:

```
No Vanity URL: http://local.cru.org:4503/content/cru/us/en/how-to-know-god.html
Vanity URL: 
  http://local.cru.org:4503/content/cru/us/en/testing.html,
  http://local.cru.org:4503/testing 
(vanityUrl = /content/cru/us/en/testing, 
 path = /content/cru/us/en/how-to-know-god)
```

I need to make sure that it is correct to send it that way, rather than sending the path, but the test is accurate to what the code is doing.

mattdrees replied:
ok, once you make sure that is correct, then this PR is good to merge.

mattdrees reviewed (commented):
Ok, another logic question. :-)

This time I'm probably wrong but I want to ask anyway.

wrandall22 committed:
This commit changes how we are handling URLs quite a bit.
 - We are now using the publisher's resource resolver to determine external URLs. This is because the sling maps are not on the author environment. This requires a change to `aem_osgi_config` as well, to add the endpoint URL.
 - We are now handling all of the vanity URLs on a page rather than just the one returned by `page.getVanityUrl()`. Some pages in production have 4+ vanity URLs on them.
 - We are now handling the `sling:redirect` property. When this is set to true, users cannot end up on the vanity URL, only on the resolved path URL. For this reason, we just send the page path to the externalizer for processing in this case.

wrandall22 committed:
Require configuration on this component. This will prevent it from being active on the publishers, where we do not want it.

wrandall22 commented:
OSGI Config changes: https://github.com/CruGlobal/aem_osgi_config/pull/255

mattdrees commented on core/src/main/java/org/cru/contentscoring/core/servlets/ResourceUrlMapperServlet.java:
```diff
@@ -0,0 +1,77 @@
+        final ResourceResolver resourceResolver) {
+
+        String pathsString = pathsParameter.getString().replace("[", "").replace("]", "");
+        String[] paths = pathsString.split(",");
```
So, you're assuming that paths will never have brackets or commas in them. Is that a safe assumption?

mattdrees replied:
You might instead consider using `getRequestParameters()`, and send a `path` query param for each path, instead of all of them in one `paths` param.

mattdrees commented on core/src/main/java/org/cru/contentscoring/core/servlets/ResourceUrlMapperServlet.java:
```diff
@@ -0,0 +1,77 @@
+        ObjectMapper objectMapper = new ObjectMapper();
+        String json = objectMapper.writeValueAsString(urls);
+        response.getWriter().write(json);
+    }
```
I think you ought to specify the Content-Type as application/json.

wrandall22 replied:
Thanks, that was the missing ingredient in being able to use the `JacksonJsonProvider`.

mattdrees commented on core/src/main/java/org/cru/contentscoring/core/webservices/SetMessageBodyReader.java:
```diff
@@ -0,0 +1,88 @@
+            return new HashSet<>();
+        }
+    }
+}
```
So, this custom reader works, but I'm not sure it's better than using `JacksonJsonProvider`, which can read a json doc into a Set<String> natively.

mattdrees replied:
In fact, except for `testShouldNotBeReadableBecauseDifferentGeneric()`, the jackson provider can be swapped in and passes `SetMessageBodyReaderTest`. I'll add a commit that does this, but you can drop it if you have a good reason for not using the jackson provider.

wrandall22 replied:
The issue was that it wasn't coming back as json, but as `application/octet-stream`

mattdrees committed:
Simplify the new URL handling logic

(from two commits back)

mattdrees committed:
Use JacksonJsonProvider instead of a custom one

I am not sure how useful the test is, now, but this commit keeps it.

mattdrees commented on core/src/main/java/org/cru/contentscoring/core/servlets/ResourceUrlMapperServlet.java:
```diff
@@ -0,0 +1,77 @@
+    methods = "GET",
+    paths = "/bin/cru/url/mapper"
+)
+public class ResourceUrlMapperServlet extends SlingSafeMethodsServlet {
```
Every time we write a new "/bin/..." servlet I cringe a little bit. I wish we could use jax-rs resource classes instead. I asked Nestor about setting this up one time, and he thought it would be possible, but we never took it beyond that.
Do you think it'd be worth our time to look into that? Not for this PR, probably.

wrandall22 replied:
Here's one solution that has arisen in that space: [caravan-jaxrs](https://github.com/wcm-io-caravan/caravan-jaxrs)

mattdrees reviewed (approved):
I made some simplifications, and I replaced the custom jax-rs reader with a OOTB jackson one. Hopefully it'll work in AEM; I didn't try it. (I verified that the maven build passes fine.)

(By the way, it's kind of annoying that the tests can't pass without having a `AEM_CONTENT_SCORING_API_KEY` env var present.)


Anyway, if you're happy with my commits then I am good with this PR, but feel free to tweak them if not.


Edit: I forgot about the Content-Type comment. I'll add a commit for that too.

mattdrees committed:
Specify Content-Type header

wrandall22 committed:
Use multiple `path` query parameters instead of one `paths` query parameter so we do not need to parse out commas and brackets.

wrandall22 committed:
Remove unused import

wrandall22 commented:
There is a `KeyService` in `giveaem` which we could extract out to the commons project and use for all of these AEM projects. That would allow us to mock it for tests.

mattdrees reviewed (commented):
👍
